tools/wake-format/formatter.h

```diff
@@ -0,0 +1,321 @@
+/*
+ * Copyright 2022 SiFive, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You should have received a copy of LICENSE.Apache2 along with
+ * this software. If not, you may obtain a copy at
+ *
+ *    https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <wcl/doc.h>
+
+#include <bitset>
+#include <cassert>
+
+#include "parser/cst.h"
+#include "parser/parser.h"
+
+#define ALWAYS_INLINE inline __attribute__((always_inline))
+
+struct ctx_t {
+  size_t width = 0;
+  size_t nest_level = 0;
+
+  ctx_t nest() {
+    ctx_t copy = *this;
+    copy.nest_level++;
+    return copy;
+  }
+
+  ctx_t sub(const wcl::doc_builder& builder) {
+    ctx_t copy = *this;
+    if (builder.has_newline()) {
+      copy.width = builder.last_width();
+    } else {
+      copy.width += builder.last_width();
+    }
+    return copy;
+  }
+};
+
+inline void space(wcl::doc_builder& builder, uint8_t count) {
+  for (uint8_t i = 0; i < count; i++) {
+    builder.append(" ");
+  }
+}
+
+struct ConsumeWhitespaceAction {
+  ALWAYS_INLINE void run(wcl::doc_builder& builder, ctx_t ctx, CSTElement& node) {
```

Comment:
oh hey thanks for adding these!

Reply:
ye, might double check that I have the macro right. I haven't inlined all that much stuff before